File: java/src/main/java/com/android/frame/mvvm/CommonBaseActivity.java
package com.android.frame.mvvm;

import android.view.LayoutInflater;

import androidx.lifecycle.ViewModelProvider;
import androidx.viewbinding.ViewBinding;

import java.lang.reflect.Method;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * Created by xuzhb on 2021/8/9
 * Desc:不需要额外声明ViewModel的Activity的父类
 */
public abstract class CommonBaseActivity<VB extends ViewBinding> extends BaseActivity<VB, BaseViewModel> implements IBaseView {

    @Override
    protected void initViewBindingAndViewModel() {
        Type superclass = getClass().getGenericSuperclass();
        Class<VB> vbClass = (Class<VB>) ((ParameterizedType) superclass).getActualTypeArguments()[0];
        try {
            Method method = vbClass.getDeclaredMethod("inflate", LayoutInflater.class);
            binding = (VB) method.invoke(null, getLayoutInflater());
            viewModel = new ViewModelProvider(this).get(BaseViewModel.class);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}

File: kotlin/src/main/java/com/android/frame/mvvm/CommonBaseActivity.kt
package com.android.frame.mvvm

import android.view.LayoutInflater
import androidx.lifecycle.ViewModelProvider
import androidx.viewbinding.ViewBinding
import java.lang.reflect.ParameterizedType

/**
 * Created by xuzhb on 2021/8/6
 * Desc:不需要额外声明ViewModel的Activity的父类
 */
abstract class CommonBaseActivity<VB : ViewBinding> : BaseActivity<VB, BaseViewModel<VB>>(), IBaseView {

    override fun initViewBindingAndViewModel() {
        val superclass = javaClass.genericSuperclass
        val vbClass = (superclass as ParameterizedType).actualTypeArguments[0] as Class<VB>
        val method = vbClass.getDeclaredMethod("inflate", LayoutInflater::class.java)
        binding = method.invoke(null, layoutInflater) as VB
        viewModel = ViewModelProvider(this).get((BaseViewModel<VB>())::class.java)
        viewModel.bind(binding)
        viewModel.observe(this, this)
    }

}